### ISSUE TITLE
Enable batch processing in container

### ISSUE THREAD

**Issue by timothyjlaurent** [kermitt2/grobid#362]
The `grobid-core/build/libs/grobid-core-`<current version>`.onejar.jar` isn't added to the docker container. It would be nice if it were so the batch mode could be run from the container.

**Comment by lfoppiano** [comment]
Hi @timothyjlaurent we could add it, indeed. 

I'm wondering why would be convenient to run the batch in the container? You could run it from the source from the release, and you would have to still get your files to the container itself. 

What do you think?

**Comment by timothyjlaurent** [comment]
Hi @lfoppiano thanks for the response.

The reason I want to run batch from the container rather than the source release is that I want to use the container in workflow that uses aws batch. Each worker container will pull a subset of the documents from s3 to a local folder and then then run batch mode on that folder, when done, the resulting xml and images will be synced back to s3. So it really makes sense to have a batch mode containerized somehow to avoid having to rebuild on each node. 

I forked the repo and hacked out a solution for my needs, https://hub.docker.com/r/invitae/grobid/, but it would be nice to have this functionality in the main repo.

**Comment by kermitt2** [comment]
Hello @timothyjlaurent !

Maybe it does not make sense for your scenario, but have you consider using the server mode of the docker containers with one of the existing GROBID clients? 
Using the service is much more efficient because it allows to use multithreading in a safe manner (with a pool of threads) while the batch is only single threaded. 

The list of available clients (they all process a list of PDF simialrly as the grobid batch mode): 
https://grobid.readthedocs.io/en/latest/Grobid-service/#clients-for-grobid-web-services

Of course we could make the grobid batch mode multithreaded quite easily and we could add batch usage to the docker container, but that would take some times.

**Comment by timothyjlaurent** [comment]
Yes I've looked into the server mode with the grobid clients - thanks for the suggestion. It works nicely on my machine, but you lose the extracted images which I'm interested in. It is also easy for me to restrict my workers to only having 1 cpu, so I can just run a couple hundred containers and have them all single processing their workload. 

The idea of having batch mode just being a chosen client calling the service for multithreading is a good idea. It would be great though to be able to get the images as well, I wonder if the server api could send back a zip file with all of this information?

**Comment by kermitt2** [comment]
> I wonder if the server api could send back a zip file with all of this information?

This is already available, just by replacing the service `processFulltextDocument` by `processFulltextAssetDocument`:

```bash
lopez@work:~/tmp/in$ curl --form input=@./020C2FBB3A8426B56756229614573FAB07D5F5AC.pdf localhost:8070/api/processFulltextAssetDocument > out.zip
lopez@work:~/tmp/in$ unzip -l out.zip 
Archive:  out.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
    50258  2018-12-15 01:22   tei.xml
   147164  2018-12-15 01:22   image-2.png
   137948  2018-12-15 01:22   image-1.png
---------                     -------
   335370                     3 files
```

It's risky because some crazy PDF can contain several ten thousand embedded images (one line images or one pixel images). There is now a default limit to 5000 extracted embedded images. It's why I don't include it in the documentation for the moment, thinking about a more intelligent way to capture images in PDF.

**Comment by lfoppiano** [comment]
I close this. @timothyjlaurent feel free to reopen should you have more questions.

**Comment by kermitt2** [comment]
I reopen because there is no satisfactory solution for the moment. 
See also #183, #243 
Opened option is also multithreaded batch processing - making sense I think, and simple to implement.
Do we want an independent batch mode or simply use the service behind a batch mode. Somehow a decision has to be made on this :) 
In any cases, I think using batch commands with the container makes sense if a batch mode is available in the application.

**Comment by lfoppiano** [comment]
ok I've added the fat jar in /opt/grobid: 

```
root@ac7f2352fe0a:/opt/grobid# java -jar grobid-core-onejar.jar 
HELP GROBID
-h: displays help
-gH: gives the path to grobid home directory.
-dIn: gives the path to the directory where the files to be processed are located, to be used only when the called method needs it.
-dOut: gives the path to the directory where the result files will be saved. The default output directory is the curent directory.
-s: is the parameter used for process using string as input and not file.
-r: recursive directory processing, default processing is not recursive.
-ignoreAssets: do not extract and save the PDF assets (bitmaps, vector graphics), by default the assets are extracted and saved.
-teiCoordinates: output a subset of the identified structures with coordinates in the original PDF, by default no coordinates are present.
-exe: gives the command to execute. The value should be one of these:
	[processDate, processAuthorsHeader, processAuthorsCitation, processAffiliation, processRawReference, processReferences, createTraining, createTrainingMonograph, createTrainingBlank, createTrainingCitationPatent, processCitationPatentST36, processCitationPatentTXT, processCitationPatentPDF, processPDFAnnotation, processHeader, processFullText, close]

root@ac7f2352fe0a:/opt/grobid# 
```

It can be tested using the snapshot image `lfoppiano/grobid:0.5.4-SNAPSHOT`

**Comment by de-code** [comment]
Slightly off topic but related, for training purpose it would also be good to have the corresponding jar available as well as I wouldn't want to have to compile everything to do the training. But on the other hand, the server image should be kept small. It might be best to push a separate server / api image from another full image or in fact separate images for `api`, `cli` and `trainer`. This could all come out of the same Docker build (just some extensions where each copy the jars from the _builder_ image).

**Comment by de-code** [comment]
I can see the discussion regarding training might get longer, moved it to #371

Would still be good to keep the `api` container as small as possible I think.
